In [1]:
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from datetime import datetime
import seaborn as sns
%matplotlib inline
import wrangle as w
import os
import env
import outliers as o
import explore as e
import functions as f

In [2]:
df = w.prep_log_data()

In [3]:
df.head()

time                                 path  user_id  cohort_id  \
date                                                                            
2018-01-26  09:55:03                                    /        1          8   
2018-01-26  09:56:02                              java-ii        1          8   
2018-01-26  09:56:05  java-ii/object-oriented-programming        1          8   
2018-01-26  09:56:06   slides/object_oriented_programming        1          8   
2018-01-26  09:56:24            javascript-i/conditionals        2         22   

                      ip  id     name     slack start_date   end_date  \
date                                                                    
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61  22    Teddy    #teddy 2018-01-08 2018-05-17   

                    created_at          updated_at  program_id  
date                                                            
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2018-01-08 13:59:10 2018-01-08 13:59:10           2

##  1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

#### Program 1

In [4]:
f.print_top_lessons1(df)

Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 1681
---------------------------
Lesson Path: index.html
Total Traffic: 1011
---------------------------
Lesson Path: javascript-i
Total Traffic: 736
---------------------------
Lesson Path: html-css
Total Traffic: 542
---------------------------


#### For program 1 the most visited lesson was javascript-1

#### Program 2

In [5]:
f.print_top_lessons2(df)

Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 35814
---------------------------
Lesson Path: javascript-i
Total Traffic: 17457
---------------------------
Lesson Path: toc
Total Traffic: 17428
---------------------------
Lesson Path: search/search_index.json
Total Traffic: 15212
---------------------------


#### For program 2 the most visited lesson was javascript-1

#### Program 3

In [6]:
f.print_top_lessons3(df)

Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 8358
---------------------------
Lesson Path: search/search_index.json
Total Traffic: 2203
---------------------------
Lesson Path: classification/overview
Total Traffic: 1785
---------------------------
Lesson Path: 1-fundamentals/modern-data-scientist.jpg
Total Traffic: 1655
---------------------------


#### For program 3 the most visited lesson was classification/overview

#### Program 4

In [7]:
f.print_top_lessons4(df)

Top 4 most visited lessons:
Lesson Path: content/html-css
Total Traffic: 2
---------------------------
Lesson Path: /
Total Traffic: 1
---------------------------
Lesson Path: content/html-css/gitbook/images/favicon.ico
Total Traffic: 1
---------------------------
Lesson Path: content/html-css/introduction.html
Total Traffic: 1
---------------------------


#### For program 4 the most visited lesson was content/html-css

## 2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?

### Acquire and Prep

In [8]:
df = w.get_log_data()
df = w.prep_log_data()

### Take cohort #28 out as they are Staff and remove cohort null values

In [9]:
df = df[df.cohort_id != 28.0]
df = df[df.cohort_id.notnull()]

### I am grouping by cohort_id and path, counting the amount of hits by cohort for each path

In [10]:
df_combined = df.groupby(['cohort_id', 'path'], as_index=False).size()
df_combined

cohort_id                                         path  size
0              1                                            /   626
1              1                  0_Classification_Algorithms     1
2              1                              2.0_Intro_Stats     1
3              1                         5.0_Intro_Regression     2
4              1                         7.0_Intro_Clustering     2
...          ...                                          ...   ...
12156        139                                       jquery     4
12157        139  jquery/essential-methods/attributes-and-css     1
12158        139                     search/search_index.json     9
12159        139                                       spring     2
12160        139                                          toc   101

[12161 rows x 3 columns]

### This function loops through each path and does the following:
* Masks the combined dataframe for the path
* Calculates the mean, max, standard deviation and upper/lower bounds of how many times the cohorts visited this page
* Finds outliers at the 3 standard deviation level
* Further limits the list to catch only one outlier (representing one single cohort going to the page while all others don't visit that much)
* prints out the scatterplot of cohort_id and visits for each page
* Returns list of paths to be zipped with the cohort list below

In [11]:
list_of_paths = f.find_outliers(df_combined) 

AttributeError: module 'functions' has no attribute 'find_outliers'

### Manual entry of the cohort numbers for MVP

In [ ]:
list_of_cohorts = [132,33,135,61,61,31,22,13,13,13,17,13,13,13,13,13,13,13,13,31,33,14,51,22,24,24,24,56,22,22,22,24,24,32,135,62,57]

### Zip the list_of_paths with list_of_cohorts and shows as a dataframe

In [ ]:
cohort_website = zip(list_of_cohorts, list_of_paths)
zipped_list = pd.DataFrame(list(cohort_website)).rename(columns={0:'Cohort Id', 1:'Website'})
zipped_list

### The above dataframe shows the cohort and path combinations found

## 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

Yes. The following students did not have many spikes in access for the curriculum after visualizing their usage: 13, 16, 59, 69, 70, 76, 90, 93, 94, 101, 102, 103, 104, 106, 109, 115, 139, 147, 156, 166, 168, 170, 178, 179, 210, 213, 214, 257, 313, 316, 371, 459, 461, 486, 513, 539, 573, 610, 611, 620, 647, 650, 778, 779, 817, 855, 894

The activity for these users tended to follow a pattern of a large spike of initial access, a delay of several months, and then another spike of activity. Likely, these students downloaded what they needed at first and then re-downloaded again after a curriculum update or when starting a new job. 

In [ ]:
# Creates a dataframe of anomalies given weight of recent values and span. This produces usage for all users with unsusual activity.
span = 30
weight = 3

anomalies = pd.DataFrame()
for u in list(df.user_id.unique()):
    user_df = e.find_anomalies(df, u, span, weight, plot=True)
    anomalies = pd.concat([anomalies, user_df], axis=0)

## 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

Yes. 78 unique users have no affiliated Codeup data. This indicates that these are users accessing curriculum who are not affiliated with Codeup. 

After creating a dataframe of access with no affiliated Codeup data, we broke down this dataframe by amount of access by IP. There were 12 IPs who accessed Codeup data in over 1000 events. The top IP, however, seems to be associated with both users with affiliated Codeup data and those without, indicating that this IP may not be suspicious.

In [ ]:
# creates separate dataframe that includes null values for analyis
df_2 = w.prep_log_data2()

In [ ]:
# creates dataframe of all user access events where the user has no affiliated data
df_sus = df_2[df_2['id'].isna()]

In [ ]:
# creates list of unique users with no affiliated data (not part of Codeup)
df_sus.user_id.unique()

In [ ]:
# shows list of ips with highest access to curriculum with no Codeup affiliated data
df_sus.ip.value_counts().head(15)

In [ ]:
# shows that this particular IP is associated with both suspicious and not suspicious users
df_2.loc[df_2['ip'] == '97.105.19.58']

## 7.  Which lessons are least accessed?

In [ ]:
lessons = df[["program_id", "cohort_id", "path"]]

result_dataframe = f.process_lessons_data(lessons)

In [ ]:
result_dataframe

## Looking through the lessons with the least traffic which was 1, there were lessons that no longer exist in the current curriculum and had testing titles such as asdfasdf as the lesson name. This tells me that there may at some point have been testing website pages. I would need a current website paths to ensure I am not delivering lessons that were "least accessed" that were never technically a lesson to began with. Using the current data, here are the 5 least visited lessons

- nlp/2-regular-expressions
- nlp/3-acquire
- nlp/exercises
- distributed-ml/3-getting-started
- distributed-ml/4-acquire

*** There are 185 more lessons with a traffic of 1, this is just 5 ***